permission/permission.go

```diff
@@ -0,0 +1,759 @@
+		}
+	}
+}
+
+// Monitors node management events and updates cache accordingly
+func (p *PermissionCtrl) manageNodePermissions() {
+	chNodeApproved := make(chan *pbind.NodeManagerNodeApproved, 1)
+	chNodeProposed := make(chan *pbind.NodeManagerNodeProposed, 1)
+	chNodeDeactivated := make(chan *pbind.NodeManagerNodeDeactivated, 1)
+	chNodeActivated := make(chan *pbind.NodeManagerNodeActivated, 1)
+	chNodeBlacklisted := make(chan *pbind.NodeManagerNodeBlacklisted)
+
+	var evtNodeApproved *pbind.NodeManagerNodeApproved
+	var evtNodeProposed *pbind.NodeManagerNodeProposed
+	var evtNodeDeactivated *pbind.NodeManagerNodeDeactivated
+	var evtNodeActivated *pbind.NodeManagerNodeActivated
+	var evtNodeBlacklisted *pbind.NodeManagerNodeBlacklisted
+
+	opts := &bind.WatchOpts{}
+	var blockNumber uint64 = 1
+	opts.Start = &blockNumber
+
+	if _, err := p.permNode.NodeManagerFilterer.WatchNodeApproved(opts, chNodeApproved); err != nil {
+		log.Info("Failed WatchNodeApproved", "error", err)
+	}
+
+	if _, err := p.permNode.NodeManagerFilterer.WatchNodeProposed(opts, chNodeProposed); err != nil {
+		log.Info("Failed WatchNodeProposed", "error", err)
+	}
+
+	if _, err := p.permNode.NodeManagerFilterer.WatchNodeDeactivated(opts, chNodeDeactivated); err != nil {
+		log.Info("Failed NodeDeactivated", "error", err)
+	}
+	if _, err := p.permNode.NodeManagerFilterer.WatchNodeActivated(opts, chNodeActivated); err != nil {
+		log.Info("Failed WatchNodeActivated", "error", err)
+	}
+
+	if _, err := p.permNode.NodeManagerFilterer.WatchNodeBlacklisted(opts, chNodeBlacklisted); err != nil {
+		log.Info("Failed NodeBlacklisting", "error", err)
+	}
+
+	for {
+		select {
+		case evtNodeApproved = <-chNodeApproved:
+			p.updatePermissionedNodes(evtNodeApproved.EnodeId, NodeAdd)
+			types.NodeInfoMap.UpsertNode(evtNodeApproved.OrgId, evtNodeApproved.EnodeId, types.NodeApproved)
+
+		case evtNodeProposed = <-chNodeProposed:
+			types.NodeInfoMap.UpsertNode(evtNodeProposed.OrgId, evtNodeProposed.EnodeId, types.NodePendingApproval)
+
+		case evtNodeDeactivated = <-chNodeDeactivated:
+			p.updatePermissionedNodes(evtNodeDeactivated.EnodeId, NodeDelete)
+			types.NodeInfoMap.UpsertNode(evtNodeDeactivated.OrgId, evtNodeDeactivated.EnodeId, types.NodeDeactivated)
+
+		case evtNodeActivated = <-chNodeActivated:
+			p.updatePermissionedNodes(evtNodeActivated.EnodeId, NodeAdd)
+			types.NodeInfoMap.UpsertNode(evtNodeActivated.OrgId, evtNodeActivated.EnodeId, types.NodeApproved)
+
+		case evtNodeBlacklisted = <-chNodeBlacklisted:
+			p.updatePermissionedNodes(evtNodeBlacklisted.EnodeId, NodeDelete)
+			p.updateDisallowedNodes(evtNodeBlacklisted.EnodeId)
+			types.NodeInfoMap.UpsertNode(evtNodeBlacklisted.OrgId, evtNodeBlacklisted.EnodeId, types.NodeBlackListed)
+		}
+	}
+}
+
+// updates node information in the permissioned-nodes.json file based on node
+// management activities in smart contract
+func (p *PermissionCtrl) updatePermissionedNodes(enodeId string, operation NodeOperation) {
```

Comment:
there might be some optimization can be done in this method:
- if there's no change, no need to write to the file
- use json streaming APIs: `json.NewEncoder`, `json.NewDecoder` rather storing data in the memory

What would be the expected behaviour when node is removed from network but writing to `permissioned-nodes.json` fails?

Reply:
Added a condition to check if file is changing or not. If not it will return. Assuming that the write failed for one of the nodes in the network, when the disconnected node is brought back:
- if it is in raft consensus - it will not be able to join the raft cluster and hence no issues
- in case of IBFT, the entire network other than this one node where write failed will reject the P2P connection and as such the node should not be able to transact. 
I have not done the changes NewEncoder/NewDecoder as the file size even in large networks will be relatively small.